Add support of custom routes to peregrine Router

Starotitorov opened the pull request:
<!-- (REQUIRED) What is the nature of this PR? -->

## This PR is a:

[ ] New feature
[x] Enhancement/Optimization
[ ] Refactor
[ ] Bugfix
[ ] Test for existing code
[ ] Documentation

<!-- (REQUIRED) What does this PR change? -->

## Summary

Add support of custom router to peregrine Router.

@zetlen, @jimbo we need to discuss the solution. I faced with some issues when tried to integrate my proposal with a new routing implementation. In this pull request I changed routing behavior a little bit, instead of rendering header, footer and 'Loading...' inside main tag, I render 'Loading...' instead of the whole page. We need to discuss the routing in the application in general to unblock EPAM team, after that I will be able to proceed with the implementation.

## Additional information

<!--
Thank you for your contribution!

Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento-research/pwa-studio/blob/master/.github/CONTRIBUTION.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PR's that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR.
-->

zetlen commented:
@Starotitorov Thank you for separating this pull request back out. I have a few questions.

1. In this implementation, the entire DOM is removed and redrawn with every navigation. Our recent routing change was meant to solve that problem, but this reintroduces it. There is a need to have persistent UI, such as header and footer. How would your system accommodate that?
2. Can you describe in numbered steps how a developer would add a custom route using this system?

Starotitorov commented:
@zetlen, Thank you for the review! You know, it is an initial implementation, I added it to implement an acceptance criteria with redirecting user to create account page from order confirmation page, initially it was easily integrated to our previous routing implementation, but now I know, that I also need to support new changes in the routing, I will add this support soon. 

Answering on the first question, yeah, I know this, I added a description right for this purpose. So, I need to clarify several things before continuing the implementation.
1. Are you OK with client side routing?
2. Are you OK with an idea of adding renderRoutes function to peregrine Router?

If the answer is yes, I will continue working on this pull request.

Answering on the second question, developer just need to implement custom renderRoutes function and pass it to peregrine Router.

Starotitorov commented:
@zetlen, @jimbo, please, take a look. I finished working on this, in this pull request I supported all changes made in the latest navigation redesign

jimbo commented on packages/peregrine/src/Router/MagentoRouteHandler.js:
```diff
@@ -11,7 +11,7 @@ const mountedInstances = new WeakSet();
     static propTypes = {
         apiBase: string.isRequired,
-        children: func,
+        renderRoutingError: func,
```
This change shouldn't be necessary.

Starotitorov replied:
In my opinion, it is a more descriptive name for the parameter. For me it was unclear why when we have an errorState we render children.

jimbo replied:
It's called [function as a child (FAAC)](https://reactjs.org/docs/render-props.html#using-props-other-than-render), and it's a common pattern.

Starotitorov replied:
yeah, I know, but why do we pass renderError function as a child, usually when I pass function as a child, I expect that I will choose what to render inside this function. so I expect that this function will be called in every case, not only in case of error.

jimbo commented on packages/peregrine/src/Router/Router.js:
```diff
@@ -1,33 +1,49 @@
-        using: func // e.g., BrowserRouter, MemoryRouter
+        using: func, // e.g., BrowserRouter, MemoryRouter
+        renderRoutes: func,
+        renderMagentoRoutingError: func
```
The purpose of the routing updates we made was to not expand the API of this router. This is counter to that purpose.

jimbo commented on packages/peregrine/src/Router/Router.js:
```diff
@@ -1,33 +1,49 @@
-                    {routeProps => (
-                        <Provider value={{ apiBase, ...routeProps }}>
-                            {children}
-                        </Provider>
```
The purpose of the recent routing changes was to provide the route info as context, which avoids render blocking issues and allows us to render route-based views at any level of the app. This would completely undo that work.

Starotitorov replied:
Official react router documentation provides a solution for the issue of blocked updates https://reacttraining.com/react-router/core/guides/redux-integration/blocked-updates and I added it. In my opinion, context api increased complexity of the code and it is possible to implement the same thing easier.

Starotitorov replied:
@jimbo What do you think?

jimbo replied:
I disagree. Context doesn't increase the complexity of the code, and it certainly doesn't need to be removed here.

Starotitorov commented:
I am going to close this pull request with routing proposal.
